services/cronback-project-srv/project_store.rs

```diff
@@ -0,0 +1,164 @@
+use async_trait::async_trait;
+use chrono::Utc;
+use lib::database::{Database, DatabaseError};
+use lib::prelude::ValidShardedId;
+use lib::types::ProjectId;
+use sea_orm::{ActiveModelTrait, EntityTrait, Set};
+
+use crate::db_model::{projects, Project, ProjectStatus, Projects};
+
+pub type ProjectStoreError = DatabaseError;
+
+#[async_trait]
+pub trait ProjectStore {
+    async fn store_project(
+        &self,
+        project: Project,
+    ) -> Result<(), ProjectStoreError>;
+
+    async fn set_status(
+        &self,
+        id: &ValidShardedId<ProjectId>,
+        status: ProjectStatus,
+    ) -> Result<(), ProjectStoreError>;
+
+    async fn get_status(
+        &self,
+        id: &ValidShardedId<ProjectId>,
+    ) -> Result<Option<ProjectStatus>, ProjectStoreError>;
+
+    async fn exists(
+        &self,
+        id: &ValidShardedId<ProjectId>,
+    ) -> Result<bool, ProjectStoreError>;
+}
+
+pub struct SqlProjectStore {
+    db: Database,
+}
+
+impl SqlProjectStore {
+    pub fn new(db: Database) -> Self {
+        Self { db }
+    }
+}
+
+#[async_trait]
+impl ProjectStore for SqlProjectStore {
+    async fn store_project(
+        &self,
+        project: Project,
+    ) -> Result<(), ProjectStoreError> {
+        let active_model: projects::ActiveModel = project.into();
+        active_model.insert(&self.db.orm).await?;
+        Ok(())
+    }
+
+    async fn set_status(
+        &self,
+        id: &ValidShardedId<ProjectId>,
+        status: ProjectStatus,
+    ) -> Result<(), ProjectStoreError> {
+        let active_model = projects::ActiveModel {
+            id: Set(id.clone()),
+            status: Set(status),
+            last_status_changed_at: Set(Utc::now()),
+            ..Default::default()
+        };
+
+        active_model.update(&self.db.orm).await?;
+        Ok(())
+    }
+
+    async fn get_status(
+        &self,
+        id: &ValidShardedId<ProjectId>,
+    ) -> Result<Option<ProjectStatus>, ProjectStoreError> {
+        Ok(Projects::find_by_id(id.clone())
+            .one(&self.db.orm)
+            .await?
+            .map(|p| p.status))
+    }
+
+    async fn exists(
+        &self,
+        id: &ValidShardedId<ProjectId>,
+    ) -> Result<bool, ProjectStoreError> {
+        Ok(self.get_status(id).await?.is_some())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use chrono::{Timelike, Utc};
+    use lib::database::{Database, DatabaseError};
+    use lib::types::ProjectId;
+
+    use super::{ProjectStore, SqlProjectStore};
+    use crate::db_model::{Project, ProjectStatus};
+
+    fn build_project(status: ProjectStatus) -> Project {
+        // Serialization drops nanoseconds, so to let's zero it here for easier
+        // equality comparisons
```

Comment:
I'm not sure I get the benefit. What do we expect when we compare created_at of different projects?

Reply:
Copypasta from other stores where we compare the values